### DISCUSSION TITLE
Table building

### DISCUSSION THREAD

**Discussion by MattBystrin** [ArthurSonzogni/FTXUI#212] [Q&A] [answered]
Hello @ArthurSonzogni ,

I'm trying to make some kind of interactive table, in which you can select the row. By now I've done something like this
+-------+-------+-------+
| vbox | vbox | Menu |
+-------+-------+--------+
But the thing is what I want the entire row have inverted style when corresponding Menu item selected. And I want to save "vbox structure" in case to save table capability to resize. How can I do that ? 
Thanks !

**Comment by ArthurSonzogni** [comment, marked as answer]
I guess you have a `std::vector<Element>` representing a column somewhere.

What about in an imperative way:

```cpp
column[selected_line] = std::move(column[selected_line]) | inverted;
```

**Reply by MattBystrin** [reply]
Thanks! I'll give a try and reply later.

**Reply by MattBystrin** [reply]
Hello, @ArthurSonzogni . I've tried to make a component that behaves as I write below. So here is my piece of code 
~~~cpp
    TableBase(std::vector<Elements> lines) : lines_(std::move(lines)) {};
    Element Render() override {
        for (auto line : lines_[selected])
            line = std::move(line) | inverted;
        return gridbox(lines_);
    }
~~~
The problem is that elements didn't become inverted. What I am doing wrong ?

**Reply by ArthurSonzogni** [reply]
I think what you wrote should work. I have no good explanation.

**Reply by tchaikov** [reply]
> Hello, @ArthurSonzogni . I've tried to make a component that behaves as I write below. So here is my piece of code 
> 
> ~~~cpp
> 
>     TableBase(std::vector<Elements> lines) : lines_(std::move(lines)) {};
> 
>     Element Render() override {
> 
>         for (auto line : lines_[selected])

Probably should capture the elements by reference?
Like
```c++
for (auto& line : lines)
```
> 
>             line = std::move(line) | inverted;
> 
>         return gridbox(lines_);
> 
>     }
> 
> ~~~
> 
> The problem is that elements didn't become inverted. What I am doing wrong ?

**Reply by ArthurSonzogni** [reply]
@tchaikov  You are right.

I would correct @MattBystrin code to:

~~~cpp
    TableBase(std::vector<Elements> lines) : lines_(std::move(lines)) {};
    Element Render() override {
        auto line = lines_;
        if (selected_ >= 0 && selected < lines.size())
          line[selected] = line[selected] | inverted;
        return gridbox(std::move(lines))
    }
~~~

**Reply by tchaikov** [reply]
yeah, thank you @ArthurSonzogni ! it's much better and even more performant, and correct =). actually, i am also trying to prototype an interactive table -- consider replicate the the pull request list UI on GitHub =)

was wondering which would the best way to structure the interactive list (or even a tree view!): a customized component rendered using a`Table` or a customized `Menu`. but i think you guys just pointed me to the right direction: to stuff the lines in a `gridbox`!

**Reply by ArthurSonzogni** [reply]
Hi @tchaikov. Here is a potential useful example for you.

I recently allowed a `Table` to contain other `Element` instead of simple `strings`.
Then I used in in [json-tui ](https://github.com/ArthurSonzogni/json-tui) to have some "interactive" table component:
![img](https://user-images.githubusercontent.com/4759106/149676534-e73b2bcc-1eef-4dde-901e-8381e9ca8f87.gif)

See:
https://github.com/ArthurSonzogni/json-tui/blob/main/src/main_ui.cpp#L276-L376+

For now, the navigation inside is using a Container::Vertical of Container::Horizontal(s). This works, but I might want to create a Container::Grid in the future to have "perfect" navigation behavior using arrow keys.

**Reply by tchaikov** [reply]
wow! that's exactly what i am after! thank you again! BTW, i created a tiny PR against json-tui: https://github.com/ArthurSonzogni/json-tui/pull/20 .
